### PR TITLE
New port for mailio.

### LINKED ISSUE #16695: [New Port Request] <mailio>
Library name: mailio

Library description: "mailio is a cross platform C++ library for MIME format and SMTP, POP3 and IMAP protocols. It is based on the standard C++ 17 and Boost library."

Source repository URL: https://github.com/karastojko/mailio

Project homepage (if different from the source repository):

Anything else that is useful to know when adding (such as optional features the library may have that should be included):

### PR DESCRIPTION
**Pull request by karastojko**
**Describe the pull request**

- What does your PR fix?
  Fixes #16695

- Which triplets are supported/not supported? Have you updated the CI baseline?
  MSVC, Gcc, Clang are supported on Linux, MacOS, Windows.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes

### CODE REVIEW

**Comment by autoantwort** [comment]
You have to format your `vcpkg.json` file with `./vcpkg format-manifest ports/mailio/vcpkg.json`. If you have committed this, you have to run `./vcpkg x-add-version mailio`

**Review by JackBoosY — CHANGES_REQUESTED** [review]
1. Please run `./vcpkg format-manifest --all` first,
2. then run `./vcpkg x-add-version --all` and commit changes.

File: ports/mailio/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "mailio",
+  "version-string": "version_0-20-0",
+  "description": "mailio is a cross platform C++ library for MIME format and SMTP, POP3 and IMAP protocols. It is based on the standard C++ 17 and Boost library.",
+  "dependencies": ["boost",	"openssl"]
```
Comment (JackBoosY) [inline ports/mailio/vcpkg.json:5]: Please add `maintainers`, see [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/specifications/manifests.md#specification).

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by JackBoosY** [comment]
```
Starting package 263/274: mailio:x64-windows-static
Building package mailio[core]:x64-windows-static...
-- Downloading https://github.com/karastojko/mailio/archive/1872f29d88a52f06cd96b611673b2e46c6b9832b.tar.gz -> karastojko-mailio-1872f29d88a52f06cd96b611673b2e46c6b9832b.tar.gz...
-- Extracting source D:/downloads/karastojko-mailio-1872f29d88a52f06cd96b611673b2e46c6b9832b.tar.gz
-- Using source at D:/buildtrees/mailio/src/46c6b9832b-9aa8aa669f.clean
-- Configuring x64-windows-static
-- Building x64-windows-static-dbg
-- Building x64-windows-static-rel
-- Installing: D:/packages/mailio_x64-windows-static/share/mailio/copyright
-- Performing post-build validation
DLLs should not be present in a static build, but the following DLLs were found:

    D:/packages/mailio_x64-windows-static/bin/mailio.dll
    D:/packages/mailio_x64-windows-static/debug/bin/mailio.dll
```
It seems that `BUILD_SHARED_LIBS` is not used in the internal cmake code.

File: ports/mailio/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        -DMAILIO_BUILD_SHARED_LIBRARY=${BUILD_SHARED}
+        -DMAILIO_BUILD_DOCUMENTATION=OFF
+        -DMAILIO_BUILD_EXAMPLES=OFF
+)
+vcpkg_cmake_install()
+
```
Comment (JackBoosY) [inline ports/mailio/portfile.cmake:20]:
```suggestion
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JackBoosY — APPROVED** [review]

File: versions/m-/mailio.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
+  "versions": [
+    {
+      "git-tree": "52e9f6449fd3bc017e7ca9a41afb9cf07a44b911",
```
Comment (JackBoosY) [inline versions/m-/mailio.json:4]:
```suggestion
      "git-tree": "125456f1ee58cb9a193ab5dfe95644150a437f0c",
```

File: ports/mailio/vcpkg.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "mailio",
+  "version-string": "version_0-20-0",
```
Comment (ras0219-msft) [inline ports/mailio/vcpkg.json:3]:
```suggestion
  "version": "0.20.0",
```

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
LGTM once the version is changed to `0.20.0`. Note that you will need to undo the changes to `versions/` (`git checkout upstream/master -- versions`) and re-run `vcpkg x-add-version mailio`.

**Comment by karastojko** [comment]
Tag renamed to 0.20.0 to follow the convention.

**Comment by ras0219-msft** [comment]
Awesome, thanks for the great library and PR :)

**Comment by autoantwort** [comment]
This port fails in the pipeline for windows uwp, as you can see [here](https://github.com/microsoft/vcpkg/pull/16788)

**Comment by kotori2** [comment]
Any idea on how to link this library? Simply use `target_link_libraries(${PROJECT_NAME} mailio)` won't work and `find_package` won't work as well...

**Comment by karastojko** [comment]
> Any idea on how to link this library? Simply use `target_link_libraries(${PROJECT_NAME} mailio)` won't work and `find_package` won't work as well...

This works for me:
`include_directories(${CMAKE_SOURCE_DIR}/libs/mailio/include)
find_library(MAILIO_LIB mailio HINTS ${CMAKE_SOURCE_DIR}/libs/mailio/build)
add_executable(${PROJ} ${PROJSRC} ${PROJINC})
target_link_libraries(${PROJ} ${Boost_LIBRARIES} Threads::Threads ${OPENSSL_LIBRARIES} ${MAILIO_LIB})
`

**Pull request merged**